<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-99.9.9">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Douglas Bates and Claudia Solis-Lemus">
<meta name="dcterms.date" content="2022-07-10">

<title>Julia Workshop - Session 2a: Data Tables and Arrow files</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Julia Workshop</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html">About</a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/crsl4/julia-workshop"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-packages-to-be-used" id="toc-load-packages-to-be-used" class="nav-link active" data-scroll-target="#load-packages-to-be-used">Load packages to be used</a></li>
  <li><a href="#objectives" id="toc-objectives" class="nav-link" data-scroll-target="#objectives">Objectives</a></li>
  <li><a href="#task-and-sample-data" id="toc-task-and-sample-data" class="nav-link" data-scroll-target="#task-and-sample-data">Task and sample data</a></li>
  <li><a href="#downloading-the-sample-data" id="toc-downloading-the-sample-data" class="nav-link" data-scroll-target="#downloading-the-sample-data">Downloading the sample data</a></li>
  <li><a href="#initial-processing" id="toc-initial-processing" class="nav-link" data-scroll-target="#initial-processing">Initial processing</a></li>
  <li><a href="#arrow-file-format" id="toc-arrow-file-format" class="nav-link" data-scroll-target="#arrow-file-format">Arrow file format</a></li>
  <li><a href="#reading-arrow-files-in-julia" id="toc-reading-arrow-files-in-julia" class="nav-link" data-scroll-target="#reading-arrow-files-in-julia">Reading Arrow files in Julia</a></li>
  <li><a href="#reading-arrow-files-in-r" id="toc-reading-arrow-files-in-r" class="nav-link" data-scroll-target="#reading-arrow-files-in-r">Reading Arrow files in R</a></li>
  <li><a href="#reading-arrow-files-in-python" id="toc-reading-arrow-files-in-python" class="nav-link" data-scroll-target="#reading-arrow-files-in-python">Reading Arrow files in Python</a></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions">Conclusions</a></li>
  <li><a href="#version-information" id="toc-version-information" class="nav-link" data-scroll-target="#version-information">Version information</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/crsl4/julia-workshop/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Session 2a: Data Tables and Arrow files</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Douglas Bates and Claudia Solis-Lemus </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">2022-07-10</p>
    </div>
  </div>
    
  </div>
  

</header>

<section id="load-packages-to-be-used" class="level2">
<h2 class="anchored" data-anchor-id="load-packages-to-be-used">Load packages to be used</h2>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Arrow             </span><span class="co"># Arrow storage and file format</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CategoricalArrays </span><span class="co"># similar to the factor type in R</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV               </span><span class="co"># read/write CSV and similar formats</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Downloads         </span><span class="co"># file downloads</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DataFrames        </span><span class="co"># versatile tabular data format</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">GZip              </span><span class="co"># utilities for compressed files</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">RCall             </span><span class="co"># run R within Julia</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Tar               </span><span class="co"># tar archive utilities</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<ul>
<li><code>Downloads</code>, <code>Gzip</code> and <code>Tar</code> are included just to demonstrate downloading and extracting tar files within Julia.</li>
<li>You may find it easier to simply click on the download link and extract the files by hand.</li>
<li>Downloading and extracting within Julia, as shown here, has a greater chance of working across various operating systems and environments in a workshop like this.</li>
<li><code>RCall</code> is included to show the use of other systems running within Julia.</li>
<li>You can instead use your favorite environment, such as <a href="https://jupyter.org">jupyterlab</a> or <a href="https://www.rstudio.com/products/rstudio/">RStudio</a>, to run Python or R</li>
<li>Note that the <a href="https://quarto.org">quarto notebooks</a> for these notes are easily converted, e.g.&nbsp;<code>quarto convert notebookname.qmd</code>, to Jupyter notebooks.</li>
</ul>
</section>
<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives</h2>
<ul>
<li>Use an example of computing interval overlaps to introduce Julia facilities for working with tabular data.</li>
<li>Introduce the Arrow format for tabular data and demonstrate its use in Julia, Python/Pandas and R.</li>
<li>Show a naive approach to computing overlaps.</li>
<li>Modify the approach to use <a href="https://github.com/BioJulia/IntervalTrees.jl">IntervalTrees.jl</a> and multi-threading.</li>
<li>Roll our own interval tree representation and use that.</li>
</ul>
<p>This notebook covers the first two objectives.</p>
</section>
<section id="task-and-sample-data" class="level2">
<h2 class="anchored" data-anchor-id="task-and-sample-data">Task and sample data</h2>
<ul>
<li><a href="https://liheng.org">Li Heng</a> provides benchmark code and sample data for comparing programming languages on Bioinformatics tasks in the <a href="https://github.com/lh3/biofast">biofast</a> repository.</li>
<li>One of these tasks, an <a href="https://bedtools.readthedocs.io/en/latest/content/tools/coverage.html">interval query</a>, takes two <a href="https://en.wikipedia.org/wiki/BED_(file_format)">.bed files</a> to compare.</li>
<li>One file, <code>ex-anno.bed</code>, contains a reference set of intervals; the other, <code>ex-rna.bed</code>, contains target intervals.</li>
<li>For each target interval, determine which reference intervals overlap with it.</li>
<li>In the benchmark both the number of reference intervals that overlap with a target and the proportion of the target covered by the overlap are computed.</li>
<li>Note that the calculation of the proportion of overlap must allow for overlapping intervals in the reference set, as shown in this <a href="https://bedtools.readthedocs.io/en/latest/content/tools/coverage.html#default-behavior">figure</a></li>
</ul>
</section>
<section id="downloading-the-sample-data" class="level2">
<h2 class="anchored" data-anchor-id="downloading-the-sample-data">Downloading the sample data</h2>
<ul>
<li>The data sets for the benchmark are available in the compressed archive <code>biofast-data-v1.tar.gz</code> at <a href="https://github.com/lh3/biofast/releases/tag/biofast-data-v1">biofast-data-v1</a>.</li>
<li>The following code chunks download the tarball, if necessary, and extract the two bedfiles to a directory <code>biofast-data-v1</code>, if necessary. (Alternatively, you can just click on the link to the tarball in the github tag page and extract the files by hand.)</li>
</ul>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>datadir <span class="op">=</span> <span class="st">"biofast-data-v1"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>tarball <span class="op">=</span> <span class="st">"</span><span class="sc">$</span>datadir<span class="st">.tar.gz"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isfile</span>(tarball)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  dataurl <span class="op">=</span> <span class="fu">joinpath</span>(</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://github.com/lh3/biofast/releases/download"</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    datadir,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    tarball,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="bu">Downloads</span>.<span class="fu">download</span>(dataurl, tarball)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(<span class="ss">`ls -lh $tarball`</span>);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-rw-rw-r-- 1 bates bates 541M Jun 16 13:34 biofast-data-v1.tar.gz</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>An expression like <code>"$datadir.tar.gz"</code> interpolates the value of <code>datadir</code> into the string, producing the name shown in the output.</p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Redundant commas are allowed at the end of the list of arguments before the <code>)</code> in a function call.</li>
</ul>
</div>
</div>
<ul>
<li>The <code>.tar.gz</code> file is about 0.5 GB. but most of that is the data for the FASTQ parsing test.</li>
<li>Extract the two <code>.bed</code> files to be used in this example, if not already present.</li>
</ul>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">isdir</span>(datadir) <span class="op">||</span> <span class="fu">mkdir</span>(datadir)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>bedfnms <span class="op">=</span> <span class="fu">joinpath</span>.(datadir, [<span class="st">"ex-anno.bed"</span>, <span class="st">"ex-rna.bed"</span>])</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>toextract <span class="op">=</span> <span class="fu">filter</span>(!isfile, bedfnms)  <span class="co"># don't overwrite existing files</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isempty</span>(toextract)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  tmpdir <span class="op">=</span> <span class="fu">gzopen</span>(tarball, <span class="st">"r"</span>) <span class="cf">do</span> io</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    Tar.<span class="fu">extract</span>(h <span class="op">-&gt;</span> <span class="fu">in</span>(h.path, toextract), io)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">end</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> pathnm <span class="kw">in</span> toextract</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mv</span>(<span class="fu">joinpath</span>(tmpdir, pathnm), pathnm; force<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">end</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(<span class="ss">`ls -lh $datadir`</span>);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>total 273M
-rw-rw-r-- 1 bates bates 3.9M Jun 16 13:34 ex-anno.arrow
-rw-r--r-- 1 bates bates  28M Jun 16 13:34 ex-anno.bed
-rw-rw-r-- 1 bates bates  39M Jun 16 13:34 ex-rna.arrow
-rw-r--r-- 1 bates bates 203M Jun 16 13:34 ex-rna.bed</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because of the way Quarto documents are rendered, this listing may show <code>.arrow</code> files that are generated later in this document.</p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The call <code>joinpath.(datadir, ["ex-anno.bed", "ex-rna.bed"])</code> is an example of <a href="https://docs.julialang.org/en/v1/manual/functions/#man-vectorized">dot broadcasting</a></p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The expression <code>h -&gt; in(h.path, toextract)</code> defines an anonymous function, in the “stabby lambda” syntax, to be used as a predicate in <code>Tar.extract</code>.</p>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(Tar.extract)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
# 4 methods for generic function <b>extract</b>:<ul><li> extract(predicate::<b>Function</b>, tarball::<b>Union{Base.AbstractCmd, AbstractString, IO}</b>) in Tar at <a href="file:///buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/Tar/src/Tar.jl" target="_blank">/home/bates/.julia/juliaup/julia-1.8.0-rc1+0~x64/share/julia/stdlib/v1.8/Tar/src/Tar.jl:218</a></li> <li> extract(predicate::<b>Function</b>, tarball::<b>Union{Base.AbstractCmd, AbstractString, IO}</b>, dir::<b>Union{Nothing, AbstractString}</b>; <i>skeleton, copy_symlinks, set_permissions</i>) in Tar at <a href="file:///buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/Tar/src/Tar.jl" target="_blank">/home/bates/.julia/juliaup/julia-1.8.0-rc1+0~x64/share/julia/stdlib/v1.8/Tar/src/Tar.jl:218</a></li> <li> extract(tarball::<b>Union{Base.AbstractCmd, AbstractString, IO}</b>) in Tar at <a href="file:///buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/Tar/src/Tar.jl" target="_blank">/home/bates/.julia/juliaup/julia-1.8.0-rc1+0~x64/share/julia/stdlib/v1.8/Tar/src/Tar.jl:248</a></li> <li> extract(tarball::<b>Union{Base.AbstractCmd, AbstractString, IO}</b>, dir::<b>Union{Nothing, AbstractString}</b>; <i>skeleton, copy_symlinks, set_permissions</i>) in Tar at <a href="file:///buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/Tar/src/Tar.jl" target="_blank">/home/bates/.julia/juliaup/julia-1.8.0-rc1+0~x64/share/julia/stdlib/v1.8/Tar/src/Tar.jl:248</a></li> </ul>
</div>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The ‘do/end’ block is yet another way of writing an anonymous function passed as a processing argument to <code>gzopen</code>.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(gzopen)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
# 4 methods for generic function <b>gzopen</b>:<ul><li> gzopen(fname::<b>AbstractString</b>) in GZip at <a href="file:///home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl" target="_blank">/home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl:264</a></li> <li> gzopen(fname::<b>AbstractString</b>, gzmode::<b>AbstractString</b>) in GZip at <a href="file:///home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl" target="_blank">/home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl:263</a></li> <li> gzopen(fname::<b>AbstractString</b>, gzmode::<b>AbstractString</b>, gz_buf_size::<b>Integer</b>) in GZip at <a href="file:///home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl" target="_blank">/home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl:236</a></li> <li> gzopen(f::<b>Function</b>, args...) in GZip at <a href="file:///home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl" target="_blank">/home/bates/.julia/packages/GZip/JNmGn/src/GZip.jl:267</a></li> </ul>
</div>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because <code>Tar.extract</code> is conservative about overwriting files and requires that the directory into which the files are extracted be empty, we extract to a freshly-created temporary directory then move the files to the desired location.</p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>It is common for packages providing utilities to avoid name conflicts by not exporting any names from their namespace (or Module).</li>
<li>The fully qualified name, <code>Tar.extract</code>, can always be used - similar to Python naming conventions.</li>
<li>If a package exports a name, say <code>foo</code>, then after the <code>using FooPackage</code> directive, the unqualified name <code>foo</code> can be used.</li>
<li>The <code>varinfo</code> function provides a listing of the names exported by a Package (formally the package’s <code>Module</code>).</li>
<li>Compare the result of <code>varinfo(Tar)</code> with that of <code>varinfo(DataFrames)</code>.</li>
</ul>
</div>
</div>
</section>
<section id="initial-processing" class="level2">
<h2 class="anchored" data-anchor-id="initial-processing">Initial processing</h2>
<ul>
<li>The <code>.bed</code> files, in this case, are simple tab-separated-value files.</li>
<li>Each of the language implementations in the <code>biofast</code> benchmark contains code to parse lines of the <code>.bed</code> files producing a <code>String</code> and two <code>Int32</code> values.</li>
<li>Furthermore the results of these benchmarks are written out as very large text files. I don’t plan to read a several-million-line text file to check if a program is working properly.</li>
<li>Why go through this parsing of text files to create a numeric representation in each language?</li>
<li>Writing code to parse a CSV or TSV file is tedious and error prone.</li>
<li>But skilled people have put a lot of work into creating packages to do just that.</li>
<li>More importantly they have tested, debugged, and documented their packages.</li>
<li>Within the <code>CSV</code> package, the <code>CSV.read</code> function reads and parses a file and converts it to a type specified by the second argument.</li>
<li>As is common for such functions, there are many, many optional named arguments</li>
<li>We read <code>ex-anno.bed</code> and create a <code>DataFrame</code> as</li>
</ul>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>annodf <span class="op">=</span> CSV.<span class="fu">read</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">joinpath</span>(datadir, <span class="st">"ex-anno.bed"</span>),</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  DataFrame;</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  delim<span class="op">=</span><span class="ch">'\t'</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int32</span>, <span class="dt">Int32</span>],</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  header<span class="op">=</span>[<span class="st">"chromo"</span>, <span class="st">"start"</span>, <span class="st">"stop"</span>],</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">

<div class="data-frame"><p>1,193,657 rows × 3 columns</p><table class="data-frame table table-sm table-striped"><thead><tr><th></th><th>chromo</th><th>start</th><th>stop</th></tr><tr><th></th><th title="String">String</th><th title="Int32">Int32</th><th title="Int32">Int32</th></tr></thead><tbody><tr><th>1</th><td>chr1</td><td>11868</td><td>12227</td></tr><tr><th>2</th><td>chr1</td><td>12612</td><td>12721</td></tr><tr><th>3</th><td>chr1</td><td>13220</td><td>14409</td></tr><tr><th>4</th><td>chr1</td><td>12009</td><td>12057</td></tr><tr><th>5</th><td>chr1</td><td>12178</td><td>12227</td></tr><tr><th>6</th><td>chr1</td><td>12612</td><td>12697</td></tr><tr><th>7</th><td>chr1</td><td>12974</td><td>13052</td></tr><tr><th>8</th><td>chr1</td><td>13220</td><td>13374</td></tr><tr><th>9</th><td>chr1</td><td>13452</td><td>13670</td></tr><tr><th>10</th><td>chr1</td><td>29533</td><td>29570</td></tr><tr><th>11</th><td>chr1</td><td>24737</td><td>24891</td></tr><tr><th>12</th><td>chr1</td><td>18267</td><td>18366</td></tr><tr><th>13</th><td>chr1</td><td>17914</td><td>18061</td></tr><tr><th>14</th><td>chr1</td><td>17605</td><td>17742</td></tr><tr><th>15</th><td>chr1</td><td>17232</td><td>17368</td></tr><tr><th>16</th><td>chr1</td><td>16857</td><td>17055</td></tr><tr><th>17</th><td>chr1</td><td>16606</td><td>16765</td></tr><tr><th>18</th><td>chr1</td><td>15795</td><td>15947</td></tr><tr><th>19</th><td>chr1</td><td>15004</td><td>15038</td></tr><tr><th>20</th><td>chr1</td><td>14403</td><td>14501</td></tr><tr><th>21</th><td>chr1</td><td>17368</td><td>17436</td></tr><tr><th>22</th><td>chr1</td><td>29553</td><td>30039</td></tr><tr><th>23</th><td>chr1</td><td>30563</td><td>30667</td></tr><tr><th>24</th><td>chr1</td><td>30975</td><td>31097</td></tr><tr><th>25</th><td>chr1</td><td>30266</td><td>30667</td></tr><tr><th>26</th><td>chr1</td><td>30975</td><td>31109</td></tr><tr><th>27</th><td>chr1</td><td>30365</td><td>30503</td></tr><tr><th>28</th><td>chr1</td><td>35720</td><td>36081</td></tr><tr><th>29</th><td>chr1</td><td>35276</td><td>35481</td></tr><tr><th>30</th><td>chr1</td><td>34553</td><td>35174</td></tr><tr><th>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td></tr></tbody></table></div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Positional arguments must come before named arguments.</li>
<li>Optionally, the comma after the last positional argument can be replace by a semicolon, as shown above.</li>
</ul>
</div>
</div>
<ul>
<li>It turns out that both of the <code>.bed</code> files contain many duplicate rows - probably not intentionally. Eliminate the duplicates with <code>unique!</code>.</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>A function’s name ending in <code>!</code> is a hint that it is a <em>mutating</em> function, which can modify one or more of its arguments.</p>
</div>
</div>
<ul>
<li>Do make one change of replacing the tags <code>"chr1"</code> up to <code>"chr9"</code> by <code>"chr01"</code> to <code>"chr09"</code> so later sorting by these strings will also give sorted chromosome numbers.</li>
<li>We use a <code>replace!</code> method takes the object to modify and one or more <code>pair</code>s of values of the form <code>from =&gt; to</code> and generate the replacements as</li>
</ul>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>replacements <span class="op">=</span>  [<span class="st">"chr</span><span class="sc">$</span>i<span class="st">"</span> <span class="op">=&gt;</span> <span class="st">"chr0</span><span class="sc">$</span>i<span class="st">"</span> for i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">9</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>9-element Vector{Pair{String, String}}:
 "chr1" =&gt; "chr01"
 "chr2" =&gt; "chr02"
 "chr3" =&gt; "chr03"
 "chr4" =&gt; "chr04"
 "chr5" =&gt; "chr05"
 "chr6" =&gt; "chr06"
 "chr7" =&gt; "chr07"
 "chr8" =&gt; "chr08"
 "chr9" =&gt; "chr09"</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The expression on the right of the <code>=</code> is called a “comprehension”. I think of it as an inside-out loop.</p>
</div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(<span class="fu">unique</span>(<span class="fu">replace!</span>(annodf.chromo, replacements<span class="op">...</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>["chr01", "chr02", "chr03", "chr04", "chr05", "chr06", "chr07", "chrX", "chr08", "chr09", "chr11", "chr10", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr20", "chr19", "chrY", "chr22", "chr21"]</code></pre>
</div>
</div>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique!</span>(annodf)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>annodf.chromo <span class="op">=</span> <span class="fu">categorical</span>(annodf.chromo)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>annodf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">

<div class="data-frame"><p>573,806 rows × 3 columns</p><table class="data-frame table table-sm table-striped"><thead><tr><th></th><th>chromo</th><th>start</th><th>stop</th></tr><tr><th></th><th title="CategoricalValue{String, UInt32}">Cat…</th><th title="Int32">Int32</th><th title="Int32">Int32</th></tr></thead><tbody><tr><th>1</th><td>chr01</td><td>11868</td><td>12227</td></tr><tr><th>2</th><td>chr01</td><td>12612</td><td>12721</td></tr><tr><th>3</th><td>chr01</td><td>13220</td><td>14409</td></tr><tr><th>4</th><td>chr01</td><td>12009</td><td>12057</td></tr><tr><th>5</th><td>chr01</td><td>12178</td><td>12227</td></tr><tr><th>6</th><td>chr01</td><td>12612</td><td>12697</td></tr><tr><th>7</th><td>chr01</td><td>12974</td><td>13052</td></tr><tr><th>8</th><td>chr01</td><td>13220</td><td>13374</td></tr><tr><th>9</th><td>chr01</td><td>13452</td><td>13670</td></tr><tr><th>10</th><td>chr01</td><td>29533</td><td>29570</td></tr><tr><th>11</th><td>chr01</td><td>24737</td><td>24891</td></tr><tr><th>12</th><td>chr01</td><td>18267</td><td>18366</td></tr><tr><th>13</th><td>chr01</td><td>17914</td><td>18061</td></tr><tr><th>14</th><td>chr01</td><td>17605</td><td>17742</td></tr><tr><th>15</th><td>chr01</td><td>17232</td><td>17368</td></tr><tr><th>16</th><td>chr01</td><td>16857</td><td>17055</td></tr><tr><th>17</th><td>chr01</td><td>16606</td><td>16765</td></tr><tr><th>18</th><td>chr01</td><td>15795</td><td>15947</td></tr><tr><th>19</th><td>chr01</td><td>15004</td><td>15038</td></tr><tr><th>20</th><td>chr01</td><td>14403</td><td>14501</td></tr><tr><th>21</th><td>chr01</td><td>17368</td><td>17436</td></tr><tr><th>22</th><td>chr01</td><td>29553</td><td>30039</td></tr><tr><th>23</th><td>chr01</td><td>30563</td><td>30667</td></tr><tr><th>24</th><td>chr01</td><td>30975</td><td>31097</td></tr><tr><th>25</th><td>chr01</td><td>30266</td><td>30667</td></tr><tr><th>26</th><td>chr01</td><td>30975</td><td>31109</td></tr><tr><th>27</th><td>chr01</td><td>30365</td><td>30503</td></tr><tr><th>28</th><td>chr01</td><td>35720</td><td>36081</td></tr><tr><th>29</th><td>chr01</td><td>35276</td><td>35481</td></tr><tr><th>30</th><td>chr01</td><td>34553</td><td>35174</td></tr><tr><th>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td></tr></tbody></table></div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Representing the <code>chromo</code> column as a <code>CategoricalArray</code> converts the individual values to indices into a vector of <code>String</code>s, which saves on space.</p>
</div>
</div>
</section>
<section id="arrow-file-format" class="level2">
<h2 class="anchored" data-anchor-id="arrow-file-format">Arrow file format</h2>
<ul>
<li>The <a href="https://arrow.apache.org">Arrow project</a> defines a memory and file format for storing and manipulating column-oriented, static, tables (i.e.&nbsp;like data frames in R, Python/Pandas, and Julia)</li>
<li>Either ‘lz4’ or ‘zstd’ compression can be used when creating an Arrow file.</li>
<li>Metadata on the names and types of columns is automatically stored. Additional column or table metadata can be specified.</li>
</ul>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>Arrow.<span class="fu">write</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"./biofast-data-v1/ex-anno.arrow"</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  annodf;</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  compress<span class="op">=:</span>lz4,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  metadata<span class="op">=</span>[</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"url"</span> <span class="op">=&gt;</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://github.com/lh3/biofast/releases/tag/biofast-data-v1"</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  ],</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<pre><code>"./biofast-data-v1/ex-anno.arrow"</code></pre>
</div>
</div>
<ul>
<li><code>bed2arrow</code> encapsulates reading the <code>.bed</code> file, replacing the strings in the <code>chromo</code> column, reducing the data frame to unique, sorted rows, and writing the Arrow file.</li>
<li>Named arguments with defaults are used to configure the “usual” call but allow for variations.</li>
<li>The Boolean <code>overwrite</code> named argument, which defaults to <code>false</code>, controls overwriting of existing files.</li>
</ul>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> [<span class="st">"url"</span> <span class="op">=&gt;</span> <span class="st">"https://github.com/lh3/biofast/releases/tag/biofast-data-v1"</span>,]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">bed2arrow</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  fnroot<span class="op">::</span><span class="dt">AbstractString</span>;</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  datadir<span class="op">=</span>datadir,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  delim<span class="op">=</span><span class="ch">'\t'</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  header<span class="op">=</span>[<span class="op">:</span>chromo, <span class="op">:</span>start, <span class="op">:</span>stop],</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int32</span>, <span class="dt">Int32</span>],</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  compress<span class="op">=:</span>lz4,</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  metadata<span class="op">=</span>metadata,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  replacements<span class="op">=</span>replacements,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  overwrite<span class="op">::</span><span class="dt">Bool</span>=<span class="cn">false</span>,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  bednm <span class="op">=</span> <span class="fu">joinpath</span>(datadir, <span class="st">"</span><span class="sc">$</span>fnroot<span class="st">.bed"</span>)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  arrownm <span class="op">=</span> <span class="fu">joinpath</span>(datadir, <span class="st">"</span><span class="sc">$</span>fnroot<span class="st">.arrow"</span>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> overwrite <span class="op">||</span> !<span class="fu">isfile</span>(arrownm)</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> <span class="fu">unique!</span>(CSV.<span class="fu">read</span>(bednm, DataFrame; header, delim, types))</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">replace!</span>(df.chromo, replacements<span class="op">...</span>)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    df.chromo <span class="op">=</span> <span class="fu">categorical</span>(df.chromo)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>    Arrow.<span class="fu">write</span>(arrownm, df; compress, metadata)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  <span class="cf">end</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> arrownm</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>rnaarrownm <span class="op">=</span> <span class="fu">bed2arrow</span>(<span class="st">"ex-rna"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>"biofast-data-v1/ex-rna.arrow"</code></pre>
</div>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(<span class="ss">`ls -lh $datadir`</span>);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>total 273M
-rw-rw-r-- 1 bates bates 3.9M Jun 16 13:37 ex-anno.arrow
-rw-r--r-- 1 bates bates  28M Jun 16 13:34 ex-anno.bed
-rw-rw-r-- 1 bates bates  39M Jun 16 13:34 ex-rna.arrow
-rw-r--r-- 1 bates bates 203M Jun 16 13:34 ex-rna.bed</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the calls to <code>CSV.read</code> and <code>Arrow.write</code>, the semicolon after the positional arguments followed by argument names only indicates that the value passed for the named argument is the object of the same name in the current namespace. That is, <code>Arrow.write(arrownm, df; compress, metadata)</code> is equivalent to <code>Arrow.write(arrownm, df; compress=compress, metadata=metadata)</code>.</p>
</div>
</div>
</section>
<section id="reading-arrow-files-in-julia" class="level2">
<h2 class="anchored" data-anchor-id="reading-arrow-files-in-julia">Reading Arrow files in Julia</h2>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>annotbl <span class="op">=</span> Arrow.<span class="fu">Table</span>(<span class="st">"./biofast-data-v1/ex-anno.arrow"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>Arrow.Table with 573806 rows, 3 columns, and schema:
 :chromo  String
 :start   Int32
 :stop    Int32

with metadata given by a Base.ImmutableDict{String, String} with 1 entry:
  "url" =&gt; "https://github.com/lh3/biofast/releases/tag/biofast-data-v1"</code></pre>
</div>
</div>
<ul>
<li>Although the schema describes the <code>chromo</code> column as <code>String</code>s the values are dictionary encoded such that each value is represented by one byte.</li>
</ul>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">typeof</span>(annotbl.chromo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>Arrow.DictEncoded{String, Int8, Arrow.List{String, Int32, Vector{UInt8}}}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="pp">@time</span> rnatbl <span class="op">=</span> Arrow.<span class="fu">Table</span>(rnaarrownm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  0.044441 seconds (321 allocations: 77.842 MiB)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>Arrow.Table with 4685080 rows, 3 columns, and schema:
 :chromo  String
 :start   Int32
 :stop    Int32

with metadata given by a Base.ImmutableDict{String, String} with 1 entry:
  "url" =&gt; "https://github.com/lh3/biofast/releases/tag/biofast-data-v1"</code></pre>
</div>
</div>
<ul>
<li>We can use operations like split-apply-combine on these tables to summarize properties</li>
</ul>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>annogdf <span class="op">=</span> <span class="fu">groupby</span>(<span class="fu">DataFrame</span>(annotbl), <span class="op">:</span>chromo)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>rnagdf <span class="op">=</span> <span class="fu">groupby</span>(<span class="fu">DataFrame</span>(rnatbl), <span class="op">:</span>chromo)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">innerjoin</span>(</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">combine</span>(rnagdf, nrow <span class="op">=&gt;</span> <span class="op">:</span>nrna),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">combine</span>(annogdf, nrow <span class="op">=&gt;</span> <span class="op">:</span>nanno);</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  on<span class="op">=:</span>chromo,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">

<div class="data-frame"><p>24 rows × 3 columns</p><table class="data-frame table table-sm table-striped"><thead><tr><th></th><th>chromo</th><th>nrna</th><th>nanno</th></tr><tr><th></th><th title="String">String</th><th title="Int64">Int64</th><th title="Int64">Int64</th></tr></thead><tbody><tr><th>1</th><td>chr01</td><td>453114</td><td>52789</td></tr><tr><th>2</th><td>chr02</td><td>298278</td><td>42563</td></tr><tr><th>3</th><td>chr03</td><td>251858</td><td>34769</td></tr><tr><th>4</th><td>chr04</td><td>132781</td><td>22500</td></tr><tr><th>5</th><td>chr05</td><td>225898</td><td>26085</td></tr><tr><th>6</th><td>chr06</td><td>307747</td><td>24935</td></tr><tr><th>7</th><td>chr07</td><td>217038</td><td>27613</td></tr><tr><th>8</th><td>chr08</td><td>141205</td><td>21603</td></tr><tr><th>9</th><td>chr09</td><td>165181</td><td>20189</td></tr><tr><th>10</th><td>chr10</td><td>162256</td><td>20061</td></tr><tr><th>11</th><td>chr11</td><td>286584</td><td>34021</td></tr><tr><th>12</th><td>chr12</td><td>292002</td><td>33524</td></tr><tr><th>13</th><td>chr13</td><td>76779</td><td>9464</td></tr><tr><th>14</th><td>chr14</td><td>171402</td><td>19995</td></tr><tr><th>15</th><td>chr15</td><td>171736</td><td>22187</td></tr><tr><th>16</th><td>chr16</td><td>192595</td><td>28276</td></tr><tr><th>17</th><td>chr17</td><td>288348</td><td>35888</td></tr><tr><th>18</th><td>chr18</td><td>60429</td><td>10179</td></tr><tr><th>19</th><td>chr19</td><td>344775</td><td>35451</td></tr><tr><th>20</th><td>chr20</td><td>109979</td><td>12387</td></tr><tr><th>21</th><td>chr21</td><td>51720</td><td>6518</td></tr><tr><th>22</th><td>chr22</td><td>132956</td><td>12724</td></tr><tr><th>23</th><td>chrX</td><td>150330</td><td>17618</td></tr><tr><th>24</th><td>chrY</td><td>89</td><td>2467</td></tr></tbody></table></div>
</div>
</div>
<ul>
<li>In the next section we will use data from <code>chr21</code> for comparative timings, because there are small, but nontrivial, number of intervals in both the reference and target groups.</li>
</ul>
</section>
<section id="reading-arrow-files-in-r" class="level2">
<h2 class="anchored" data-anchor-id="reading-arrow-files-in-r">Reading Arrow files in R</h2>
<ul>
<li><p>In R (and in Python) the Arrow file format is confounded with an earlier file format called Feather and referred to as <code>Feather V2</code>.</p></li>
<li><p>In R the <code>arrow::read_feather</code> function returns a tibble. In an R session it looks like</p></li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(tibble)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> arrow<span class="sc">::</span><span class="fu">read_feather</span>(<span class="st">"biofast-data-v1/ex-rna.arrow"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 4,685,080 × 3</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>   chromo     start      stop</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>fct<span class="sc">&gt;</span>      <span class="er">&lt;</span>int<span class="sc">&gt;</span>     <span class="er">&lt;</span>int<span class="sc">&gt;</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> chr02  <span class="dv">216499331</span> <span class="dv">216501458</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> chr07  <span class="dv">101239611</span> <span class="dv">101245071</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> chr19   <span class="dv">49487626</span>  <span class="dv">49491841</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> chr10   <span class="dv">80155590</span>  <span class="dv">80169336</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> chr17   <span class="dv">76270411</span>  <span class="dv">76271290</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> chr06   <span class="dv">31268756</span>  <span class="dv">31272069</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> chr05  <span class="dv">170083214</span> <span class="dv">170083368</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> chr19   <span class="dv">51989731</span>  <span class="dv">51989996</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> chr18   <span class="dv">55225980</span>  <span class="dv">55226732</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> chr16   <span class="dv">84565611</span>  <span class="dv">84566066</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a><span class="co"># … with 4,685,070 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The <code>RCall</code> package in Julia allows for running an R process within a Julia session.</li>
<li>One way of executing R code with <code>RCall</code> is to prepend <code>R</code> to a string. This causes the string to be evaluated in R.</li>
<li><code>$</code>-interpolation in the string causes a Julia object to be copied into the R environment and its name in R interpolated.</li>
</ul>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>R<span class="st">"""</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="st">library(tibble)</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="st">glimpse(rnatbl &lt;- arrow::read_feather(</span><span class="sc">$</span>rnaarrownm<span class="st">))</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span>;</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 4,685,080
Columns: 3
$ chromo &lt;fct&gt; chr02, chr07, chr19, chr10, chr17, chr06, chr05, chr19, chr18, …
$ start  &lt;int&gt; 216499331, 101239611, 49487626, 80155590, 76270411, 31268756, 1…
$ stop   &lt;int&gt; 216501458, 101245071, 49491841, 80169336, 76271290, 31272069, 1…</code></pre>
</div>
</div>
</section>
<section id="reading-arrow-files-in-python" class="level2">
<h2 class="anchored" data-anchor-id="reading-arrow-files-in-python">Reading Arrow files in Python</h2>
<ul>
<li>The <code>pyarrow</code> package includes <code>pyarrow.feather</code>. Its use in a Python session looks like</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> pyarrow.feather <span class="im">as</span> fea</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> fea.read_table(<span class="st">'./biofast-data-v1/ex-rna.arrow'</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>pyarrow.Table</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>chromo: dictionary<span class="op">&lt;</span>values<span class="op">=</span>string, indices<span class="op">=</span>int8, ordered<span class="op">=</span><span class="dv">0</span><span class="op">&gt;</span> <span class="kw">not</span> null</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>start: int32 <span class="kw">not</span> null</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>stop: int32 <span class="kw">not</span> null</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="op">----</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>chromo: [  <span class="op">--</span> dictionary:</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>[<span class="st">"chr01"</span>,<span class="st">"chr02"</span>,<span class="st">"chr03"</span>,<span class="st">"chr04"</span>,<span class="st">"chr05"</span>,...,<span class="st">"chr20"</span>,<span class="st">"chr21"</span>,<span class="st">"chr22"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>]  <span class="op">--</span> indices:</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">18</span>,<span class="dv">9</span>,<span class="dv">16</span>,...,<span class="dv">16</span>,<span class="dv">15</span>,<span class="dv">10</span>,<span class="dv">22</span>,<span class="dv">0</span>]]</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>start: [[<span class="dv">216499331</span>,<span class="dv">101239611</span>,<span class="dv">49487626</span>,<span class="dv">80155590</span>,<span class="dv">76270411</span>,...,<span class="dv">7014179</span>,<span class="dv">75627747</span>,<span class="dv">59636724</span>,<span class="dv">153785767</span>,<span class="dv">182839364</span>]]</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>stop: [[<span class="dv">216501458</span>,<span class="dv">101245071</span>,<span class="dv">49491841</span>,<span class="dv">80169336</span>,<span class="dv">76271290</span>,...,<span class="dv">7014515</span>,<span class="dv">75631483</span>,<span class="dv">59666963</span>,<span class="dv">153787586</span>,<span class="dv">182887745</span>]]</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> fea.read_feather(<span class="st">'./biofast-data-v1/ex-rna.arrow'</span>)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>        chromo      start       stop</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span>        chr02  <span class="dv">216499331</span>  <span class="dv">216501458</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>        chr07  <span class="dv">101239611</span>  <span class="dv">101245071</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>        chr19   <span class="dv">49487626</span>   <span class="dv">49491841</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>        chr10   <span class="dv">80155590</span>   <span class="dv">80169336</span></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>        chr17   <span class="dv">76270411</span>   <span class="dv">76271290</span></span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>...        ...        ...        ...</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="dv">4685075</span>  chr17    <span class="dv">7014179</span>    <span class="dv">7014515</span></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="dv">4685076</span>  chr16   <span class="dv">75627747</span>   <span class="dv">75631483</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="dv">4685077</span>  chr11   <span class="dv">59636724</span>   <span class="dv">59666963</span></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a><span class="dv">4685078</span>   chrX  <span class="dv">153785767</span>  <span class="dv">153787586</span></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a><span class="dv">4685079</span>  chr01  <span class="dv">182839364</span>  <span class="dv">182887745</span></span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4685080</span> rows x <span class="dv">3</span> columns]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><code>read_table</code> returns a <code>Table</code> object, <code>read_feather</code> returns a Pandas dataframe.</p></li>
<li><p>The <code>PyCall</code> package for <code>Julia</code> starts a Python process and allows communication with it, including data transfer.</p></li>
<li><p>I use this instead of the Python REPL when working with both Julia and Python.</p></li>
<li><p>Configuring Python, Conda, pyarrow, pandas, and PyCall across platforms is sufficiently complicated to almost surely cause failures for some workshop participants. Instead of evaluating this code chunk we quote the results.</p></li>
</ul>
<div class="sourceCode" id="cb32"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>julia<span class="op">&gt;</span> <span class="im">using</span> <span class="bu">PyCall</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>julia<span class="op">&gt;</span> fea <span class="op">=</span> <span class="fu">pyimport</span>(<span class="st">"pyarrow.feather"</span>);</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>julia<span class="op">&gt;</span> fea.<span class="fu">read_table</span>(rnaarrownm)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>PyObject pyarrow.Table</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>chromo<span class="op">:</span> dictionary<span class="op">&lt;</span>values<span class="op">=</span>string, indices<span class="op">=</span>int8, ordered<span class="op">=</span><span class="fl">0</span><span class="op">&gt;</span> not null</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>start<span class="op">:</span> int32 not null</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>stop<span class="op">:</span> int32 not null</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="op">----</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>chromo<span class="op">:</span> [  <span class="op">--</span> dictionary<span class="op">:</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>[<span class="st">"chr01"</span>,<span class="st">"chr02"</span>,<span class="st">"chr03"</span>,<span class="st">"chr04"</span>,<span class="st">"chr05"</span>,<span class="op">...</span>,<span class="st">"chr20"</span>,<span class="st">"chr21"</span>,<span class="st">"chr22"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>]  <span class="op">--</span> indices<span class="op">:</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1</span>,<span class="fl">6</span>,<span class="fl">18</span>,<span class="fl">9</span>,<span class="fl">16</span>,<span class="op">...</span>,<span class="fl">16</span>,<span class="fl">15</span>,<span class="fl">10</span>,<span class="fl">22</span>,<span class="fl">0</span>]]</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>start<span class="op">:</span> [[<span class="fl">216499331</span>,<span class="fl">101239611</span>,<span class="fl">49487626</span>,<span class="fl">80155590</span>,<span class="fl">76270411</span>,<span class="op">...</span>,<span class="fl">7014179</span>,<span class="fl">75627747</span>,<span class="fl">59636724</span>,<span class="fl">153785767</span>,<span class="fl">182839364</span>]]</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>stop<span class="op">:</span> [[<span class="fl">216501458</span>,<span class="fl">101245071</span>,<span class="fl">49491841</span>,<span class="fl">80169336</span>,<span class="fl">76271290</span>,<span class="op">...</span>,<span class="fl">7014515</span>,<span class="fl">75631483</span>,<span class="fl">59666963</span>,<span class="fl">153787586</span>,<span class="fl">182887745</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="conclusions" class="level2">
<h2 class="anchored" data-anchor-id="conclusions">Conclusions</h2>
<ul>
<li>Julia provides an impressive array of tools for Bioinformatics and similar tasks</li>
<li>We have shown the use of <a href="https://github.com/apache/arrow-julia">Arrow.jl</a>, <a href="https://github.com/JuliaData/CategoricalArrays.jl">CategoricalArrays.jl</a> <a href="https://github.com/JuliaData/CSV.jl">CSV.jl</a>, and <a href="https://github.com/JuliaData/DataFrames.jl">DataFrames</a> for data input, storage and manipulation.</li>
<li>Although not shown here <a href="https://github.com/jkrumbiegel/DataFrameMacros.jl">DataFrameMacros.jl</a> (or <a href="https://github.com/JuliaData/DataFramesMeta.jl">DataFramesMeta.jl</a>) and <a href="https://github.com/jkrumbiegel/Chain.jl">Chain.jl</a> are worth considering for more sophisticated work with DataFrames.</li>
<li>A general framework for working with tabular data is provided in <a href="https://github.com/JuliaData/Tables.jl">Tables.jl</a> (not shown here).</li>
<li>We have shown the use of <a href="https://github.com/JuliaPy/PyCall.jl">PyCall.jl</a> and <a href="https://github.com/JuliaInterop/RCall.jl">RCall.jl</a> for running and communicating with other language systems within Julia.</li>
<li>We have shown the use of utility packages <a href="https://github.com/JuliaLang/Downloads.jl">Downloads.jl</a>, <a href="https://github.com/JuliaIO/GZip.jl.git">GZip.jl</a> and <a href="https://github.com/JuliaIO/Tar.jl">Tar.jl</a> for scripting, say within <a href="https://quarto.org">Quarto</a> documents like these.</li>
<li>Take a moment to look at the repositories for some of these Julia packages. Most (all?) of them are 100% Julia code.</li>
</ul>
</section>
<section id="version-information" class="level2">
<h2 class="anchored" data-anchor-id="version-information">Version information</h2>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">versioninfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Julia Version 1.8.0-rc1
Commit 6368fdc656 (2022-05-27 18:33 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 8 × 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, tigerlake)
  Threads: 4 on 8 virtual cores</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>